sloghandler.go

```diff
@@ -0,0 +1,71 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"context"
+	"log/slog"
+)
+
+type slogHandler struct {
+	sink        LogSink
+	groupPrefix string
+}
+
+func (l *slogHandler) Enabled(ctx context.Context, level slog.Level) bool {
+	return l.sink.Enabled(levelFromSlog(level))
+}
+
+func (l *slogHandler) Handle(ctx context.Context, record slog.Record) error {
```

Comment:
Somewhere we need to handle caller info.  My PR gets it right by hardcoding a depth, which I don't think slog will want to guarantee.

Reply:
Maybe we need to call Callers() again and walk back until I find the PC they gave us, and add that many frames to the logr depth?

Reply:
> My PR gets it right by hardcoding a depth, which I don't think slog will want to guarantee.

Only if the code using the handler is using `slog.Logger`. We can't assume that, the code might also use something else with different offsets.

> Maybe we need to call Callers() again and walk back until I find the PC they gave us, and add that many frames to the logr depth?

Urgh. That probably works in most usages, but not when the record was produced in one goroutine and then gets logged somewhere else entirely.

Reply:
https://github.com/go-logr/logr/issues/201